File: waf_database.js
const { MongoClient } = require("mongodb");
const GORILLA_DB_NAME = "Gorilla_WAF_DB";
const BLACK_LIST_NAME = "BlackList";
const PACKET_LIST_NAME = "IncomingPackets";
const MAX_NUM_OF_ATTACKS = 3;

class MongoDB {
    constructor (db_ip, db_port) {
        
        //Create URI by the user IP and Port
        const URI = "mongodb://" + db_ip + ":" + db_port + "/" + GORILLA_DB_NAME;

        //Create new client by the URI
        this.client = new MongoClient(URI);
        this.start();
    }

    /*
        Function that create the database and collection if they don't exist.
        input: none
        output: none
    */
    async start (){

        try {
            //Conmect the client to the server.
            await this.client.connect();

            //Create both of the collections.
            await this.client.db().createCollection(BLACK_LIST_NAME);
            await this.client.db().createCollection(PACKET_LIST_NAME);
        } catch (e) {
            console.error();
        } finally {
            //Close the database.
            await this.client.close();
        }
    }

    /*
        Add new entry to the "Blacklist" collection, including all the information needed.
        input:
            ip_address (str): Attacker's IP Address.
            num_of_attacks (int): The number of attacks that the attacker did.
            is_blocked (bool): If the attacker is blocked or not (True if blocked, False if not) [3+ attacks = blocked].

        output: none.
    */
    async add_to_BlackList (ip_address, num_of_attacks, is_blocked) {
        await this.client.db(GORILLA_DB_NAME).collection(BLACK_LIST_NAME).insertOne(
            {
                "IP_Address" : ip_address,
                "Num_of_Attacks" : num_of_attacks,
                "Is_Blocked" : is_blocked
            }
        );
    }


    /*
        Add new entry to the "IncomingPacket" collection, including all the information needed.
        input:
            client_ip (str): Source IP, Client's IP.
            client_port (int): Source port, Client's Port.
            content (str): The HTTP message (Only application level).
            is_safe (bool): True if the packet is safe, False if not (in case of attack).
            name_of_attack (str): Attack's name (If packet is safe - leave empty).

        output: none.
    */
    async add_to_IncomingPackets (client_ip, client_port, content, is_safe, name_of_attack) {
        await this.client.db(GORILLA_DB_NAME).collection(PACKET_LIST_NAME).insertOne(
            {
                "Client's_IP" : client_ip,
                "Client's_Port" : client_port,
                "Num_of_Attacks" : content ,
                "Is_Safe" : is_safe, 
                "Name_of_Attack" : name_of_attack
            }
        );
    }


    /*
        Update the number of attacks that the attacker did, in the "Blacklist" collection, by the given IP.
        input:
            ip_address (string): Attacker's IP Address, the IP address in the entry we want to find and update.
            new_val (string): New value for the number of attacks.

        output: none.
    */
    async update_BlackList_num_of_attacks (ip_address, new_val) {

        await this.client.db(GORILLA_DB_NAME).collection(BLACK_LIST_NAME).findOneAndUpdate(
            //The identifier field
            {
                "IP_Address" : ip_address
            },
            
            //The field we want to update
            {$set: 
                {
                    "Num_of_Attacks" : new_val
                } 
            }
        );
    }


    /*
        Update the 'Is Blocked' field in the "Blacklist" collection, by the given IP.
        input:
            ip_address (string): Attacker's IP Address, the IP address in the entry we want to find and update.
            new_val (bool): New value for 'Is Blocked'.

        output: none.
    */
    async update_BlackList_is_blocked (ip_address, new_val) {

        await this.client.db(GORILLA_DB_NAME).collection(BLACK_LIST_NAME).findOneAndUpdate(
            //The identifier field
            {
                "IP_Address" : ip_address
            },
            
            //The field we want to update

            {$set: 
                {
                    "Is_Blocked" : new_val
                } 
            }
        );
    }

    
    /*
        Function that print list of the user's databases.
        input: the client
        output: none
    */
    async listDatabases (client) {
        //Create list of all the databases using admin (everyone has this database in MongoDB)
        const databaseList = await client.db().admin().listDatabases();

        console.log("Databases:");

        //Print them
        databaseList.databases.forEach(db => {
            console.log(`-${db.name}`);
        });
    }


    /*
        Function that change the user information according to the attack that he preformed.
        input:
            ip_address (str): Attacker's IP Address.

        output: none
    */
    async burn_attacker (ip_address) {

        //Create an array that contains the user information.
        const all_details =  await this.client.db(GORILLA_DB_NAME).collection(BLACK_LIST_NAME).find({"IP_Address" : ip_address}).toArray();


        try
        {
            //If the user doesn't exist in the database then this command will return error.
            var current_num_of_attacks = all_details[0].Num_of_Attacks;

            //Check if the user reached the maximum number of attacks, if he did, then we change his status
            if (current_num_of_attacks == MAX_NUM_OF_ATTACKS) {
                this.update_BlackList_is_blocked(ip_address, true);
            }

            //If the user hasn't reached the maximum number of attacks we add one.
            else
            {
                this.update_BlackList_num_of_attacks(ip_address, current_num_of_attacks + 1)
            }

        }
        catch
        {
            //If the user doesn't exist, then we add him to the database.
            this.add_to_BlackList(ip_address, 1, false);
        }
    }
}
// To connect to another file.
module.exports = MongoDB;


//const c = new MongoDB ("localhost", "27017");
//c.add_to_BlackList(1, 2, true);
//c.add_to_IncomingPackets(4444, 22, "Tttttt", false, "DDOS");
//c.update_BlackList_is_blocked(5555,"mememee");
//c.burn_attacker(2);
//console.log(c.is_in_BlackList(1));
//c.is_in_BlackList(1);
//console.log(c.is_in_BlackList(1));

File: scan_software/waf_database.py
import pymongo

GORILLA_DB_NAME = "Gorilla_WAF_DB"
MAX_NUM_OF_ATTACKS = 3

class MongoDB:

    def __init__(self, db_ip, db_port):
        """
        Constructor that creates a connection to the database, and creates the collections if they don't exist.
        
        Args:
            db_ip (str): The IP address of the database.
            db_port (int): The port of the database.
        """
        
        # Open the connection with the MongoDB's IP&Port and init the '__db' field:
        self.__client = pymongo.MongoClient(db_ip, db_port)
        self.__db = self.__client[GORILLA_DB_NAME]
        
        # Try creating "Blacklist" and "Incoming Requests" collections (if not exist):
        try:
            self.__db.create_collection("Blacklist")
        except pymongo.errors.CollectionInvalid:
            pass
        
        try:
            self.__db.create_collection("IncomingRequests")
        except pymongo.errors.CollectionInvalid:
            pass

    
    def add_to_blacklist(self, ip_address, name_of_attack, num_of_attacks, is_blocked):
        """
        Add new entry to the "Blacklist" collection, including all the information needed.
        
        Args:
            ip_address (str): Attacker's IP Address.
            name_of_attack (str): Attack's name.
            num_of_attacks (int): The number of attacks that the attacker did.
            is_blocked (bool): If the attacker is blocked or not (True if blocked, False if not) [3+ attacks = blocked].
        """
        
        # Define the entry that we want to add to the collection, and add it:
        entry_to_add = {
            "IP Address": ip_address,
            "Attacks Performed": name_of_attack,
            "Num of Attacks": num_of_attacks,
            "Is Blocked": is_blocked
        }
        self.__db["Blacklist"].insert_one(entry_to_add)
        
        
    def add_to_incoming_requests(self, client_ip, client_port, content, is_safe, name_of_attack, time):
        """
        Add new entry to the "IncomingRequests" collection, including all the information needed.
        
        Args:
            client_ip (str): Source IP, Client's IP.
            client_port (int): Source port, Client's Port.
            content (str): The HTTP message (Only application level).
            is_safe (bool): True if the packet is safe, False if not (in case of attack).
            name_of_attack (str): Attack's name (If packet is safe - leave empty).
            time (str): The time that the packet was received.
        """
        
        # Define the entry that we want to add to the collection, and add it:
        entry_to_add = {
            "Client's IP": client_ip,
            "Client's Port": client_port,
            "HTTP Request": content ,
            "Is Safe": is_safe, 
            "Name of Attack": name_of_attack,
            "Time" : time
        }
        self.__db["IncomingRequests"].insert_one(entry_to_add)
        
        
    def find_in_blacklist(self, ip_address):
        """
        Return entry from "Blacklist" collection, by the given IP.
        
        Args:
            ip_address (str): Attacker's IP Address, the IP address we want to find.
        Returns:
            dictionary: The entry that we found.
        """
        
        # Define the entry we want to find in "Blacklist" collection, and return the result:
        entry_to_find = {
            "IP Address": ip_address, 
        }
        return self.__db["Blacklist"].find_one(entry_to_find)
   
   
    def find_in_incoming_packets(self, client_ip, client_port):
        """
        Return all entries from "IncomingRequests" collection, by the given IP.
        
        Args:
            client_ip (str): Client's IP Address, the IP address we want to find.
            client_port (int): Client's port, the port we want to find.

        Returns:
            _type_: dictionary - The entries that we found.
        """
        
        # Define the entry we want to find in "Incoming Packets" collection, and return the result:
        entry_to_find = {
            "Client IP": client_ip, 
            "Client Port": client_port
        }
        return self.__db["IncomingPackets"].find(entry_to_find)
    
    def get_attacks_performed(self, ip_address):
        """
        Function that returns the attacks that the user preformed.
        
        Args:
            ip_address (str): Attacker's IP Address.
        """
        return self.find_in_blacklist(ip_address)["Attacks Performed"]
    
    
    def update_blacklist(self, ip_address, num_of_attacks, attack_name):
        """
        Update the number of attacks that the attacker did, in the "Blacklist" collection, by the given IP.
        
        Args:
            ip_address (str): Attacker's IP Address, the IP address in the entry we want to find and update.
            new_val (int): New value for the number of attacks.
        """
        
        # Define the entry we want to update in "Blacklist" collection, set the new value and update it:
        entry_to_update = { 
            "IP Address": ip_address,
        }
        
        #current_attacks = self.get_attack_preformed(ip_address) + "," + attack_name
        current_attacks = self.find_in_blacklist(ip_address)["Attacks Performed"] + "," + attack_name
        
        new_value = {}
        #If the client is at the maximum number of attacks, block him:
        if num_of_attacks == MAX_NUM_OF_ATTACKS:
            new_value = { "$set": {"Attacks Performed": current_attacks, "Num of Attacks": num_of_attacks, "Is Blocked": True } }
        else:
            new_value = { "$set": {"Attacks Performed": current_attacks, "Num of Attacks": num_of_attacks } }
            
        self.__db["Blacklist"].update_one(entry_to_update, new_value)
        
    
    def update_blacklist_is_blocked(self, ip_address, new_val):
        """
        Update the 'Is Blocked' field in the "Blacklist" collection, by the given IP.
        
        Args:
            ip_address (str): Attacker's IP Address, the IP address in the entry we want to find and update.
            new_val (bool): New value for 'Is Blocked'.
        """
        
        # Define the entry we want to update in "Blacklist" collection, set the new value and update it:
        entry_to_update = { 
            "IP Address": ip_address,
        }
        new_value = { "$set": { "Is Blocked": new_val } } 
        self.__db["Blacklist"].update_one(entry_to_update, new_value)
        
        
    def delete_one_blacklist(self, ip_address):
        """
        Delete one entry from the "Blacklist" collection, by the given IP.
        Usually used when the attacker's 'Num of Attacks' = 0.

        Args:
            ip_address (str): Attacker's IP Address, the IP address in the entry we want to find and delete.
        """
        # Define the entry we want to delete in "Blacklist" collection and delete it:
        entry_to_delete = {
            "IP Address": ip_address, 
        }
        self.__db["Blacklist"].delete_one(entry_to_delete)
        
    def is_in_blacklist(self, ip_address):
        """
        Check if the given IP is in the "Blacklist" collection.
        
        Args:
            ip_address (str): Attacker's IP Address, the IP address we want to check.
        Returns:
            bool: True if the IP is in the collection, False if not.
        """
        # Define the entry we want to find in "Blacklist" collection, and return the result:
        entry_to_find = {
            "IP Address": ip_address, 
        }
        return self.__db["Blacklist"].find_one(entry_to_find) is not None
    
    
    def is_blocked(self, ip_address):
        """
            Function that checks if the user is blocked.
            Args:
                ip_address (str): Attacker's IP Address.
        """
        if (not self.is_in_blacklist(ip_address)):
            return False
        return self.find_in_blacklist(ip_address)["Is Blocked"]
        